### PR TITLE
microdds_client: add namespace to partecipant name

### PR DESCRIPTION
**Pull request by beniaminopozzan**
## Describe problem solved by this pull request

1. The `microdds_client` did not add the namespace to the participant name. As a result when working with multiple vehicles and ROS2 all microdds_agent had the same name.
2. in `sitl` the `microdds_client` was automatically started on `udp` and port `8888`. With multiple vehicles this caused having all the clients connected to the same agent with the same key identifier, breaking the bridge regardless the presence of the namespace.

## Describe your solution

1. The `participant_xml` is created taking into account the desired namespace.
2. The parameter `XRCE_DDS_KEY` is added and used to specify the session key.
3. In `sitl` the `microdds_client` is automatically started on `udp`. ~~Depending on the `px4_instance` variable the port and the namespace are selected:
`microdds_client start -t udp -p "$((px4_instance+8888))" -n "px4_${px4_instance}"`~~
If `px4_instance` is zero, no multiple clients are expected: `XRCE_DDS_KEY` is untouched and no namespace is defined.
If `px4_instance` is greater than zero, multiple clients are expected: `XRCE_DDS_KEY` is set equal to `px4_instance` and the namespace follow the convention `px4_${px4_instance}`.
4. Regardless the value of `px4_instance`, if the environment variable `PX4_MICRODDS_NS` is set, then it is used as namespace.

## Describe possible alternatives
Have the namespace in the form of a parameter (requires string parameters)

### Single client scenario
This solution has been tested in ingnition simulation. For two vehicles:
**terminal 1**
`make px4_sitl gz_x500`
**terminal 2**
`ros2 run micro_ros_agent micro_ros_agent udp4 --port 8888`

#### Expected Output on terminal 3:
```
$ ros2 node list
/px4_micro_xrce_dds
$ ros2 topic list
/fmu/in/obstacle_distance
/fmu/in/offboard_control_mode
/fmu/in/onboard_computer_status
/fmu/in/sensor_optical_flow
/fmu/in/telemetry_status
/fmu/in/trajectory_setpoint
/fmu/in/vehicle_attitude_setpoint
/fmu/in/vehicle_command
/fmu/in/vehicle_mocap_odometry
/fmu/in/vehicle_rates_setpoint
/fmu/in/vehicle_trajectory_bezier
/fmu/in/vehicle_trajectory_waypoint
/fmu/in/vehicle_visual_odometry
/fmu/out/failsafe_flags
/fmu/out/sensor_combined
/fmu/out/timesync_status
/fmu/out/vehicle_attitude
/fmu/out/vehicle_control_mode
/fmu/out/vehicle_global_position
/fmu/out/vehicle_gps_position
/fmu/out/vehicle_local_position
/fmu/out/vehicle_odometry
/fmu/out/vehicle_status
/parameter_events
/rosout
```

### Multi-client scenario
This solution has been tested in ingnition simulation:
**terminal 1**
```
make px4_sitl
PX4_SYS_AUTOSTART=4001 PX4_GZ_MODEL_POSE="0,0"  PX4_GZ_MODEL=x500 ./build/px4_sitl_default/bin/px4 -i 1
```
**terminal 2**
`PX4_SYS_AUTOSTART=4001 PX4_GZ_MODEL_POSE="0,1"  PX4_GZ_MODEL=x500 ./build/px4_sitl_default/bin/px4 -i 2`
**terminal 3**
`ros2 run micro_ros_agent micro_ros_agent udp4 --port 8888`

#### Expected Output on terminal 4:
```
$ ros2 node list
/px4_1/px4_micro_xrce_dds
/px4_2/px4_micro_xrce_dds
$ ros2 topic list
/parameter_events
/px4_1/fmu/in/obstacle_distance
/px4_1/fmu/in/offboard_control_mode
/px4_1/fmu/in/onboard_computer_status
/px4_1/fmu/in/sensor_optical_flow
/px4_1/fmu/in/telemetry_status
/px4_1/fmu/in/trajectory_setpoint
/px4_1/fmu/in/vehicle_attitude_setpoint
/px4_1/fmu/in/vehicle_command
/px4_1/fmu/in/vehicle_mocap_odometry
/px4_1/fmu/in/vehicle_rates_setpoint
/px4_1/fmu/in/vehicle_trajectory_bezier
/px4_1/fmu/in/vehicle_trajectory_waypoint
/px4_1/fmu/in/vehicle_visual_odometry
/px4_1/fmu/out/failsafe_flags
/px4_1/fmu/out/sensor_combined
/px4_1/fmu/out/timesync_status
/px4_1/fmu/out/vehicle_attitude
/px4_1/fmu/out/vehicle_control_mode
/px4_1/fmu/out/vehicle_global_position
/px4_1/fmu/out/vehicle_gps_position
/px4_1/fmu/out/vehicle_local_position
/px4_1/fmu/out/vehicle_odometry
/px4_1/fmu/out/vehicle_status
/px4_2/fmu/in/obstacle_distance
/px4_2/fmu/in/offboard_control_mode
/px4_2/fmu/in/onboard_computer_status
/px4_2/fmu/in/sensor_optical_flow
/px4_2/fmu/in/telemetry_status
/px4_2/fmu/in/trajectory_setpoint
/px4_2/fmu/in/vehicle_attitude_setpoint
/px4_2/fmu/in/vehicle_command
/px4_2/fmu/in/vehicle_mocap_odometry
/px4_2/fmu/in/vehicle_rates_setpoint
/px4_2/fmu/in/vehicle_trajectory_bezier
/px4_2/fmu/in/vehicle_trajectory_waypoint
/px4_2/fmu/in/vehicle_visual_odometry
/px4_2/fmu/out/failsafe_flags
/px4_2/fmu/out/sensor_combined
/px4_2/fmu/out/timesync_status
/px4_2/fmu/out/vehicle_attitude
/px4_2/fmu/out/vehicle_control_mode
/px4_2/fmu/out/vehicle_global_position
/px4_2/fmu/out/vehicle_gps_position
/px4_2/fmu/out/vehicle_local_position
/px4_2/fmu/out/vehicle_odometry
/px4_2/fmu/out/vehicle_status
/rosout
```

## Additional context
I used ignition fortress source compiled for this test. And even if #20319 is fully functional, as the associated PR in gz has been merged, I had to manually modify the gazebo version to overcome #20342.

### CODE REVIEW

**Comment by dagar** [comment]
Can you fix the astyle fail?. https://github.com/PX4/PX4-Autopilot/actions/runs/3301234874/jobs/5446479275
It can be skipped with a tag if you want to preserve the whitespace.

**Comment by dagar** [comment]
Wouldn't we want all instances connecting to the same agent (especially in a SITL simulation scenario)?

@spiderkeys any suggestions/thoughts here?

**Comment by beniaminopozzan** [comment]
Style fixed.

You are right, having all instances connected to the same agents would be preferable. However from what I've tested, to do so not only each instance need a unique namespace, it also need a unique key identifier, which is currently fixed to 0xAAAABBBB.
For this reason I would suggest to parameterize the session key value so that we can change it as we please, but at the same time

1. I don't know if there are specific rules to assign the session key value.
2. Currently the session key value is implemented as an `unsigned int` while the parameter would be just an `int`.

If there are no particular rules for the session key value, we could define a parameter acting as a _session index_ such that `key = 0xAAAABBBB + session_index` and the use the same index to define the namespace, like `"px4"+session_index`.

**Comment by beniaminopozzan** [comment]
I went ahead and parametrized the session key. Now multiple client can connect to the same agent.
This should cover most of the uses in simulation.

**Comment by max11gen** [comment]
Thanks, @beniaminopozzan, I just tried this and it works great. Is there a reason for this not to be merged, @dagar ?

File: ROMFS/px4fmu_common/init.d-posix/rcS [hunk 1]
```diff
@@ -260,7 +260,13 @@ fi
...
 
 # Try to start the microdds_client with UDP transport if module exists
-microdds_client start -t udp -p 8888
+if [ $px4_instance -eq 0 ]
+then
+	microdds_client start -t udp -p 8888
+else
+	param set XRCE_DDS_KEY ${px4_instance}
+	microdds_client start -t udp -p 8888 -n "px4_$px4_instance"
+fi
```
Comment (kribe48) [inline ROMFS/px4fmu_common/init.d-posix/rcS:269]: I suggest using an environment variable for the namespace, e.g., "PX4_MICRODDS_NS". This gives the user more freedom of selecting the namespace, e.g. based on vehicle. Then, the rcS would look something like this 
if [ ! -z $PX4_MICRODDS_NS ] ; then
param set XRCE_DDS_KEY ${px4_instance}
microdds_client start -t udp -p 8888 -n ${PX4_MICRODDS_NS} 
else 
microdds_client start -t udp -p 8888

What do you think of this?

**Review by dagar — APPROVED** [review]

**Pull request merged**